extensions/wikia/Email/Controller/WatchedPageController.class.php

```diff
@@ -122,38 +120,187 @@ private function getDetails() {
+		return wfMessage( 'emailext-watchedpage-view-all-changes',
+			$this->title->getFullURL( [
+				'action' => 'history'
+			] ),
+			$this->title->getPrefixedText()
+		)->inLanguage( $this->targetLang )->parse();
+	}
+
+	/**
+	 * @return Array
+	 */
+	protected function getContentFooterMessages() {
+		return [
+			$this->getArticleLinkText(),
+			$this->getAllChangesText(),
+		];
+	}
+}
+
+class WatchedPageEditedController extends WatchedPageController {
+	/**
+	 * @return String
+	 */
+	protected function getSubjectMessageKey() {
+		return 'emailext-watchedpage-article-edited-subject';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSummaryMessageKey() {
+		return 'emailext-watchedpage-article-edited';
+	}
+}
+
+class WatchedPageProtectedController extends WatchedPageController {
+	/**
+	 * @return String
+	 */
+	protected function getSubjectMessageKey() {
+		return 'emailext-watchedpage-article-protected-subject';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSummaryMessageKey() {
+		return 'emailext-watchedpage-article-protected';
+	}
+}
+
+class WatchedPageUnprotectedController extends WatchedPageController {
+	/**
+	 * @return String
+	 */
+	protected function getSubjectMessageKey() {
+		return 'emailext-watchedpage-article-unprotected-subject';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSummaryMessageKey() {
+		return 'emailext-watchedpage-article-unprotected';
+	}
+}
+
+class WatchedPageDeletedController extends WatchedPageController {
+	/**
+	 * @return String
+	 */
+	protected function getSubjectMessageKey() {
+		return 'emailext-watchedpage-article-deleted-subject';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSummaryMessageKey() {
+		return 'emailext-watchedpage-article-deleted';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getButtonTextMessageKey() {
+		return 'emailext-watchedpage-deleted-button-text';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getButtonLink() {
+		return $this->title->getFullUrl();
+	}
+
+	/**
+	 * @return Array
+	 */
+	protected function getContentFooterMessages() {
+		return [];
+	}
+}
+
+class WatchedPageRenamedController extends WatchedPageController {
+	protected $newTitle;
+
+	public function initEmail() {
+		parent::initEmail();
+
+		$this->newTitle = \WikiPage::factory( $this->title )->getRedirectTarget();
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSubjectMessageKey() {
+		return 'emailext-watchedpage-article-renamed-subject';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getSummaryMessageKey() {
+		return 'emailext-watchedpage-article-renamed';
+	}
+
+	/**
+	 * @return String
+	 */
+	protected function getArticleLinkText() {
+		return wfMessage( 'emailext-watchedpage-article-link-text',
+			$this->newTitle->getFullURL(),
+			$this->newTitle->getPrefixedText()
+		)->inLanguage( $this->targetLang )->parse();
 	}
 
 	/**
 	 * @return String
 	 */
```

Comment:
Maybe the WatchedPageController version of this could take an optional ```$title``` argument and then this could just be:
```
protected function getAllChangesText() {
    return parent::getAllChangesText( $this->newTitle );
}
```